File: nativelib/src/main/scala/scala/scalanative/posix/termios.scala
package scala.scalanative
package posix

import native.{CArray, CChar, CInt, CLong, CStruct7, Nat, Ptr, extern, name}
import posix.sys.types.pid_t

@extern
object termios {
  import Nat._

  // types

  type tcflag_t = CLong
  type cc_t     = CChar
  type speed_t  = CLong
  type NCCS     = Digit[_2, _0]
  type c_cc     = CArray[cc_t, NCCS]

  type termios = CStruct7[
    tcflag_t, /* c_iflag - input flags   */
    tcflag_t, /* c_oflag - output flags  */
    tcflag_t, /* c_cflag - control flags */
    tcflag_t, /* c_lflag - local flags   */
    c_cc, /* cc_t c_cc[NCCS] - control chars */
    speed_t, /* c_ispeed - input speed   */
    speed_t /* c_ospeed - output speed  */
  ]

  // functions

  def cfgetispeed(termios_p: Ptr[termios]): speed_t              = extern
  def cfgetospeed(termios_p: Ptr[termios]): speed_t              = extern
  def cfsetispeed(termios_p: Ptr[termios], speed: speed_t): CInt = extern
  def cfsetospeed(termios_p: Ptr[termios], speed: speed_t): CInt = extern
  def tcdrain(fd: CInt): CInt                                    = extern
  def tcflow(fd: CInt, action: CInt): CInt                       = extern
  def tcflush(fd: CInt, queueSelector: CInt): CInt               = extern
  def tcgetattr(fd: CInt, termios_p: Ptr[termios]): CInt         = extern
  def tcgetsid(i: CInt): pid_t                                   = extern
  def tcsendbreak(fd: CInt, duration: CInt): CInt                = extern
  def tcsetattr(fd: CInt,
                optionalActions: CInt,
                termios_p: Ptr[termios]): CInt = extern

  // symbolic constants for use as subscripts for the array c_cc

  @name("scalanative_termios_veof")
  def VEOF: CInt = extern
  @name("scalanative_termios_veof")
  def VEOL: CInt = extern
  @name("scalanative_termios_verase")
  def VERASE: CInt = extern
  @name("scalanative_termios_vintr")
  def VINTR: CInt = extern
  @name("scalanative_termios_vkill")
  def VKILL: CInt = extern
  @name("scalanative_termios_vmin")
  def VMIN: CInt = extern
  @name("scalanative_termios_vquit")
  def VQUIT: CInt = extern
  @name("scalanative_termios_vstart")
  def VSTART: CInt = extern
  @name("scalanative_termios_vstop")
  def VSTOP: CInt = extern
  @name("scalanative_termios_vsusp")
  def VSUSP: CInt = extern
  @name("scalanative_termios_vtime")
  def VTIME: CInt = extern

  // Input Modes - symbolic constants for use as flags in the c_iflag field

  @name("scalanative_termios_brkint")
  def BRKINT: CInt = extern
  @name("scalanative_termios_icrnl")
  def ICRNL: CInt = extern
  @name("scalanative_termios_ignbrk")
  def IGNBRK: CInt = extern
  @name("scalanative_termios_igncr")
  def IGNCR: CInt = extern
  @name("scalanative_termios_ignpar")
  def IGNPAR: CInt = extern
  @name("scalanative_termios_inlcr")
  def INLCR: CInt = extern
  @name("scalanative_termios_inpck")
  def INPCK: CInt = extern
  @name("scalanative_termios_istrip")
  def ISTRIP: CInt = extern
  @name("scalanative_termios_ixany")
  def IXANY: CInt = extern
  @name("scalanative_termios_ixoff")
  def IXOFF: CInt = extern
  @name("scalanative_termios_ixon")
  def IXON: CInt = extern
  @name("scalanative_termios_parmrk")
  def PARMRK: CInt = extern

  // Output Modes - symbolic constants for use as flags in the c_oflag field

  @name("scalanative_termios_opost")
  def OPOST: CInt = extern
  // [XSI] follows
  @name("scalanative_termios_onlcr")
  def ONLCR: CInt = extern
  @name("scalanative_termios_ocrnl")
  def OCRNL: CInt = extern
  @name("scalanative_termios_onocr")
  def ONOCR: CInt = extern
  @name("scalanative_termios_onlret")
  def ONLRET: CInt = extern
  @name("scalanative_termios_ofdel")
  def OFDEL: CInt = extern
  @name("scalanative_termios_ofill")
  def OFILL: CInt = extern
  // Masks - Linux [requires _BSD_SOURCE or _SVID_SOURCE or _XOPEN_SOURCE]
  @name("scalanative_termios_nldly")
  def NLDLY: CInt = extern
  @name("scalanative_termios_nl0")
  def NL0: CInt = extern
  @name("scalanative_termios_nl1")
  def NL1: CInt = extern
  @name("scalanative_termios_crdly")
  def CRDLY: CInt = extern
  @name("scalanative_termios_cr0")
  def CR0: CInt = extern
  @name("scalanative_termios_cr1")
  def CR1: CInt = extern
  @name("scalanative_termios_cr2")
  def CR2: CInt = extern
  @name("scalanative_termios_cr3")
  def CR3: CInt = extern
  @name("scalanative_termios_tabdly")
  def TABDLY: CInt = extern
  @name("scalanative_termios_tab0")
  def TAB0: CInt = extern
  @name("scalanative_termios_tab1")
  def TAB1: CInt = extern
  @name("scalanative_termios_tab2")
  def TAB2: CInt = extern
  @name("scalanative_termios_tab3")
  def TAB3: CInt = extern
  @name("scalanative_termios_bsdly")
  def BSDLY: CInt = extern
  @name("scalanative_termios_bs0")
  def BS0: CInt = extern
  @name("scalanative_termios_bs1")
  def BS1: CInt = extern
  @name("scalanative_termios_vtdly")
  def VTDLY: CInt = extern
  @name("scalanative_termios_vt0")
  def VT0: CInt = extern
  @name("scalanative_termios_vt1")
  def VT1: CInt = extern
  @name("scalanative_termios_ffdly")
  def FFDLY: CInt = extern
  @name("scalanative_termios_ff0")
  def FF0: CInt = extern
  @name("scalanative_termios_ff1")
  def FF1: CInt = extern

  // Baud Rate Selection - valid values for objects of type speed_t

  @name("scalanative_termios_b0")
  def B0: CInt = extern
  @name("scalanative_termios_b50")
  def B50: CInt = extern
  @name("scalanative_termios_b75")
  def B75: CInt = extern
  @name("scalanative_termios_b110")
  def B110: CInt = extern
  @name("scalanative_termios_b134")
  def B134: CInt = extern
  @name("scalanative_termios_b150")
  def B150: CInt = extern
  @name("scalanative_termios_b200")
  def B200: CInt = extern
  @name("scalanative_termios_b300")
  def B300: CInt = extern
  @name("scalanative_termios_b600")
  def B600: CInt = extern
  @name("scalanative_termios_b1200")
  def B1200: CInt = extern
  @name("scalanative_termios_b1800")
  def B1800: CInt = extern
  @name("scalanative_termios_b2400")
  def B2400: CInt = extern
  @name("scalanative_termios_b4800")
  def B4800: CInt = extern
  @name("scalanative_termios_b9600")
  def B9600: CInt = extern
  @name("scalanative_termios_b19200")
  def B19200: CInt = extern
  @name("scalanative_termios_b38400")
  def B38400: CInt = extern

  // Control Modes - symbolic constants for use as flags in the c_cflag field

  @name("scalanative_termios_csize")
  def CSIZE: CInt = extern
  @name("scalanative_termios_cs5")
  def CS5: CInt = extern
  @name("scalanative_termios_cs6")
  def CS6: CInt = extern
  @name("scalanative_termios_cs7")
  def CS7: CInt = extern
  @name("scalanative_termios_cs8")
  def CS8: CInt = extern
  @name("scalanative_termios_cstopb")
  def CSTOPB: CInt = extern
  @name("scalanative_termios_cread")
  def CREAD: CInt = extern
  @name("scalanative_termios_parenb")
  def PARENB: CInt = extern
  @name("scalanative_termios_parodd")
  def PARODD: CInt = extern
  @name("scalanative_termios_hupcl")
  def HUPCL: CInt = extern
  @name("scalanative_termios_clocal")
  def CLOCAL: CInt = extern

  // Local Modes - symbolic constants for use as flags in the c_lflag field

  @name("scalanative_termios_echo")
  def ECHO: CInt = extern
  @name("scalanative_termios_echoe")
  def ECHOE: CInt = extern
  @name("scalanative_termios_echok")
  def ECHOK: CInt = extern
  @name("scalanative_termios_echonl")
  def ECHONL: CInt = extern
  @name("scalanative_termios_icanon")
  def ICANON: CInt = extern
  @name("scalanative_termios_iexten")
  def IEXTEN: CInt = extern
  @name("scalanative_termios_isig")
  def ISIG: CInt = extern
  @name("scalanative_termios_noflsh")
  def NOFLSH: CInt = extern
  @name("scalanative_termios_tostop")
  def TOSTOP: CInt = extern

  // Attribute Selection - symbolic constants for use with tcsetattr()

  @name("scalanative_termios_tcsanow")
  def TCSANOW: CInt = extern
  @name("scalanative_termios_tcsadrain")
  def TCSADRAIN: CInt = extern
  @name("scalanative_termios_tcsaflush")
  def TCSAFLUSH: CInt = extern

  // Line Control - symbolic constants for use with tcflush()

  @name("scalanative_termios_tciflush")
  def TCIFLUSH: CInt = extern
  @name("scalanative_termios_tcioflush")
  def TCIOFLUSH: CInt = extern
  @name("scalanative_termios_tcoflush")
  def TCOFLUSH: CInt = extern

  // Line Control cont. - symbolic constants for use with tcflow()

  @name("scalanative_termios_tcioff")
  def TCIOFF: CInt = extern
  @name("scalanative_termios_tcion")
  def TCION: CInt = extern
  @name("scalanative_termios_tcooff")
  def TCOOFF: CInt = extern
  @name("scalanative_termios_tcoon")
  def TCOON: CInt = extern

}

File: nativelib/shared/src/main/resources/posix/termios.c
#include <termios.h>

// symbolic constants for use as subscripts for the array c_cc

int scalanative_termios_veof() { return VEOF; }
int scalanative_termios_veol() { return VEOL; }
int scalanative_termios_verase() { return VERASE; }
int scalanative_termios_vintr() { return VINTR; }
int scalanative_termios_vkill() { return VKILL; }
int scalanative_termios_vmin() { return VMIN; }
int scalanative_termios_vquit() { return VQUIT; }
int scalanative_termios_vstart() { return VSTART; }
int scalanative_termios_vstop() { return VSTOP; }
int scalanative_termios_vsusp() { return VSUSP; }
int scalanative_termios_vtime() { return VTIME; }

// Input Modes - symbolic constants for use as flags in the c_iflag field

int scalanative_termios_brkint() { return BRKINT; }
int scalanative_termios_icrnl() { return ICRNL; }
int scalanative_termios_ignbrk() { return IGNBRK; }
int scalanative_termios_igncr() { return IGNCR; }
int scalanative_termios_ignpar() { return IGNPAR; }
int scalanative_termios_inlcr() { return INLCR; }
int scalanative_termios_inpck() { return INPCK; }
int scalanative_termios_istrip() { return ISTRIP; }
int scalanative_termios_ixany() { return IXANY; }
int scalanative_termios_ixoff() { return IXOFF; }
int scalanative_termios_ixon() { return IXON; }
int scalanative_termios_parmrk() { return PARMRK; }

// Output Modes - symbolic constants for use as flags in the c_oflag field

int scalanative_termios_opost() { return OPOST; }
// [XSI] follows
int scalanative_termios_onlcr() { return ONLCR; }
int scalanative_termios_ocrnl() { return OCRNL; }
int scalanative_termios_onocr() { return ONOCR; }
int scalanative_termios_onlret() { return ONLRET; }
int scalanative_termios_ofdel() { return OFDEL; }
int scalanative_termios_ofill() { return OFILL; }
// Masks - Linux [requires _BSD_SOURCE or _SVID_SOURCE or _XOPEN_SOURCE]
int scalanative_termios_nldly() { return NLDLY; }
int scalanative_termios_nl0() { return NL0; }
int scalanative_termios_nl1() { return NL1; }
int scalanative_termios_crdly() { return CRDLY; }
int scalanative_termios_cr0() { return CR0; }
int scalanative_termios_cr1() { return CR1; }
int scalanative_termios_cr2() { return CR2; }
int scalanative_termios_cr3() { return CR3; }
int scalanative_termios_tabdly() { return TABDLY; }
int scalanative_termios_tab0() { return TAB0; }
int scalanative_termios_tab1() { return TAB1; }
int scalanative_termios_tab2() { return TAB2; }
int scalanative_termios_tab3() { return TAB3; }
int scalanative_termios_bsdly() { return BSDLY; }
int scalanative_termios_bs0() { return BS0; }
int scalanative_termios_bs1() { return BS1; }
int scalanative_termios_vtdly() { return VTDLY; }
int scalanative_termios_vt0() { return VT0; }
int scalanative_termios_vt1() { return VT1; }
int scalanative_termios_ffdly() { return FFDLY; }
int scalanative_termios_ff0() { return FF0; }
int scalanative_termios_ff1() { return FF1; }

// Baud Rate Selection - valid values for objects of type speed_t

int scalanative_termios_b0() { return B0; }
int scalanative_termios_b50() { return B50; }
int scalanative_termios_b75() { return B75; }
int scalanative_termios_b110() { return B110; }
int scalanative_termios_b134() { return B134; }
int scalanative_termios_b150() { return B150; }
int scalanative_termios_b200() { return B200; }
int scalanative_termios_b300() { return B300; }
int scalanative_termios_b600() { return B600; }
int scalanative_termios_b1200() { return B1200; }
int scalanative_termios_b1800() { return B1800; }
int scalanative_termios_b2400() { return B2400; }
int scalanative_termios_b4800() { return B4800; }
int scalanative_termios_b9600() { return B9600; }
int scalanative_termios_b19200() { return B19200; }
int scalanative_termios_b38400() { return B38400; }

// Control Modes - symbolic constants for use as flags in the c_cflag field

int scalanative_termios_csize() { return CSIZE; }
int scalanative_termios_cs5() { return CS5; }
int scalanative_termios_cs6() { return CS6; }
int scalanative_termios_cs7() { return CS7; }
int scalanative_termios_cs8() { return CS8; }
int scalanative_termios_cstopb() { return CSTOPB; }
int scalanative_termios_cread() { return CREAD; }
int scalanative_termios_parenb() { return PARENB; }
int scalanative_termios_parodd() { return PARODD; }
int scalanative_termios_hupcl() { return HUPCL; }
int scalanative_termios_clocal() { return CLOCAL; }

// Local Modes - symbolic constants for use as flags in the c_lflag field

int scalanative_termios_echo() { return ECHO; }
int scalanative_termios_echoe() { return ECHOE; }
int scalanative_termios_echok() { return ECHOK; }
int scalanative_termios_echonl() { return ECHONL; }
int scalanative_termios_icanon() { return ICANON; }
int scalanative_termios_iexten() { return IEXTEN; }
int scalanative_termios_isig() { return ISIG; }
int scalanative_termios_noflsh() { return NOFLSH; }
int scalanative_termios_tostop() { return TOSTOP; }

// Attribute Selection - symbolic constants for use with tcsetattr()

int scalanative_termios_tcsanow() { return TCSANOW; }
int scalanative_termios_tcsadrain() { return TCSADRAIN; }
int scalanative_termios_tcsaflush() { return TCSAFLUSH; }

// Line Control - symbolic constants for use with tcflush()

int scalanative_termios_tciflush() { return TCIFLUSH; }
int scalanative_termios_tcioflush() { return TCIOFLUSH; }
int scalanative_termios_tcoflush() { return TCOFLUSH; }

// Line Control cont. - symbolic constants for use with tcflow()

int scalanative_termios_tcioff() { return TCIOFF; }
int scalanative_termios_tcion() { return TCION; }
int scalanative_termios_tcooff() { return TCOOFF; }
int scalanative_termios_tcoon() { return TCOON; }
